File: android/core/models/src/main/kotlin/com/nexussoft/verbum/models/ReadingActivity.kt
package com.nexussoft.verbum.models

import java.time.Instant
import java.time.ZoneId

/** Device-local chapter visits, not a count of completed chapters. */
data class ReadingActivity(val visits: List<Visit> = emptyList(), val days: Set<String> = emptySet()) {
    data class Visit(val reference: PassageReference, val lastOpened: Long)
    fun record(reference: PassageReference, at: Instant, zone: ZoneId): ReadingActivity {
        val book = BibleBook.book(reference.bookId) ?: return this
        if (reference.chapter !in 1..book.chapterCount) return this
        val chapter = PassageReference(reference.bookId, reference.chapter)
        return copy(
            visits = listOf(Visit(chapter, at.toEpochMilli())) + visits.filter { it.reference != chapter },
            days = days + at.atZone(zone).toLocalDate().toString(),
        )
    }
}

File: VerbumKit/Sources/Models/ReadingActivity.swift
import Foundation

/// Device-local activity. Opening a chapter is not a claim that it was completed.
public struct ReadingActivity: Codable, Equatable, Sendable {
    public struct Visit: Codable, Equatable, Sendable, Identifiable {
        public var reference: PassageReference
        public var lastOpened: Date
        public var id: String { ReaderCanon.key(reference) }
    }
    public private(set) var visits: [Visit] = []
    public private(set) var days: Set<String> = []
    public init() {}

    public mutating func record(_ reference: PassageReference, at date: Date, calendar: Calendar) {
        guard let book = BibleBook.book(id: reference.bookId), (1...book.chapterCount).contains(reference.chapter) else { return }
        let chapter = PassageReference(bookId: reference.bookId, chapter: reference.chapter)
        visits.removeAll { $0.reference == chapter }
        visits.insert(Visit(reference: chapter, lastOpened: date), at: 0)
        let parts = calendar.dateComponents([.year, .month, .day], from: date)
        days.insert("\(parts.year ?? 0)-\(parts.month ?? 0)-\(parts.day ?? 0)")
    }
}
